source/common/upstream/conn_pool_map_impl.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
```

Comment:
I'm not sure how you feel about this method of templating. I like it, because it has similar semantics to a header/cpp file. Only the class using the template needs  to know about the details of its implementation. I can put this all into one header if you prefer.

Reply:
This LGTM.